File: src/core/self_auditor.py
import os
import re
import json
from datetime import datetime

class SelfAuditor:
    """
    KALI Self-Auditor (Python Core)
    Enforces 'Codebase Purification' before training cycles.
    """
    def __init__(self):
        self.scan_dirs = [
            os.path.join("src", "core"),
            os.path.join("modules"),
            os.path.join("src", "static", "js")
        ]
        self.rules = [
            {
                "name": "HARDCODED_PATTERN",
                "regex": r"(https?://[^\s'\"]+)|(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})",
                "severity": "high",
                "message": "Hardcoded external IP or URL detected. Move to Sovereign Registry (config.json)."
            },
            {
                "name": "MOCK_SYSTEM_FALLBACK",
                "regex": r"return\s+.*mock.*|callback\(.*mock.*\)|res\.send\(.*mock.*\)|simulation_mode",
                "severity": "critical",
                "message": "Active mock fallback or simulation string detected."
            }
        ]
        self.exclude_files = ["config.json", "config_manager.py", "audit_report.json"]

    def run_audit(self):
        print("[AUDITOR] Initiating codebase purification scan...")
        report = {"timestamp": datetime.now().isoformat(), "issues": []}

        for directory in self.scan_dirs:
            if os.path.exists(directory):
                self._scan_directory(directory, report)

        # Persistence
        report_path = os.path.join("data", "audit_report.json")
        with open(report_path, "w") as f:
            json.dump(report, f, indent=4)
            
        return report

    def _scan_directory(self, directory, report):
        for root, _, files in os.walk(directory):
            for file in files:
                if file.endswith(('.js', '.py')):
                    self._analyze_file(os.path.join(root, file), report)

    def _analyze_file(self, file_path, report):
        # 🔱 ARCHITECTURAL PROTECTION: Exclude core configuration and the auditor itself
        excluded = ["config.json", "self_auditor.py", ".claude", "logs", "data", "scratch"]
        if any(ex in file_path for ex in excluded):
            return True
            
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                lines = f.readlines()
                for idx, line in enumerate(lines):
                    for rule in self.rules:
                        if re.search(rule["regex"], line, re.IGNORECASE):
                            report["issues"].append({
                                "file": os.path.relpath(file_path),
                                "line": idx + 1,
                                "rule": rule["name"],
                                "severity": rule["severity"],
                                "message": rule["message"],
                                "snippet": line.strip()
                            })
        except Exception as e:
            print(f"[AUDITOR] Failed to analyze {file_path}: {e}")

File: src/core/self_auditor.js
/**
 * KALI Self-Auditor — v1.0 SOVEREIGN
 * Scans the codebase for drift, hardcoding, and inefficiencies.
 */

const fs = require('fs');
const path = require('path');

class SelfAuditor {
    constructor() {
        this.scanDirs = [
            path.join(__dirname, '../../src/core'),
            path.join(__dirname, '../../src/static/js')
        ];
        this.rules = [
            {
                name: 'HARDCODED_PATTERN',
                regex: /(https?:\/\/(?!127\.0\.0\.1|localhost)[^\s'"]+)|(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}(?!\.0\.0\.1))/,
                severity: 'high',
                message: 'Hardcoded external IP or URL detected. Must use configuration registry.'
            },
            {
                name: 'MOCK_SYSTEM_FALLBACK',
                regex: /return\s+.*mock.*|callback\(.*mock.*\)|res\.send\(.*mock.*\)/i,
                severity: 'critical',
                message: 'Active mock fallback detected. System must be grounded in real swarm telemetry.'
            },
            {
                name: 'DUPLICATE_LOGIC_CANDIDATE',
                regex: /function\s+\w+\(\)\s*\{\s*\w+\(\);?\s*\}/,
                severity: 'medium',
                message: 'Potential passthrough/redundant function detected. Candidate for architectural flattening.'
            },
            {
                name: 'UNSTABLE_IMPORT',
                regex: /require\(.*(?:\.\.\/){3,}.*\)/,
                severity: 'medium',
                message: 'Deep relative import detected (>3 levels). Suggests fragmented architecture.'
            },
            {
                name: 'DEAD_CODE_MARKER',
                regex: /\/\/\s*TEMP:|\/\/\s*FIXME:|\/\/\s*STUB:/i,
                severity: 'low',
                message: 'Development marker remains in sovereign core.'
            }
        ];
    }

    runAudit() {
        console.log("[SelfAuditor] Initiating zero-drift codebase scan...");
        const report = { timestamp: new Date().toISOString(), issues: [] };

        this.scanDirs.forEach(dir => {
            if (!fs.existsSync(dir)) return;
            this._scanDirectory(dir, report);
        });

        // Write report
        const reportPath = path.join(__dirname, '../../data/audit_report.json');
        fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));

        if (report.issues.length > 0) {
            console.warn(`[SelfAuditor] Warning: ${report.issues.length} structural issues detected. Check data/audit_report.json`);
        } else {
            console.log("[SelfAuditor] Audit complete: Codebase is 100% pure. Zero drift.");
        }

        return report;
    }

    _scanDirectory(dir, report) {
        const files = fs.readdirSync(dir);
        files.forEach(file => {
            const fullPath = path.join(dir, file);
            const stat = fs.statSync(fullPath);
            if (stat.isDirectory()) {
                this._scanDirectory(fullPath, report);
            } else if (file.endsWith('.js') || file.endsWith('.py')) {
                this._analyzeFile(fullPath, report);
            }
        });
    }

    _analyzeFile(filePath, report) {
        // 🔱 ARCHITECTURAL PROTECTION: Exclude core configuration and the auditor itself
        const excluded = ["config.json", "self_auditor.js", "self_auditor.py", ".claude", "logs", "data", "scratch"];
        if (excluded.some(ex => filePath.includes(ex))) return;

        const content = fs.readFileSync(filePath, 'utf8');
        const lines = content.split('\n');

        lines.forEach((line, idx) => {
            this.rules.forEach(rule => {
                if (rule.regex.test(line)) {
                    report.issues.push({
                        file: path.basename(filePath),
                        line: idx + 1,
                        rule: rule.name,
                        severity: rule.severity,
                        message: rule.message,
                        snippet: line.trim()
                    });
                }
            });
        });
    }
}

module.exports = new SelfAuditor();
